tests/qnn/test_keras.py

```diff
@@ -0,0 +1,450 @@
+        taken with respect to the trainable variables"""
+        c, w = get_circuit
+        layer = KerasLayer(c, w, output_dim)
+        x = tf.ones((1, n_qubits))
+
+        with tf.GradientTape() as tape:
+            out_layer = layer(x)
+
+        g_layer = tape.gradient(out_layer, layer.trainable_variables)
+
+        with tf.GradientTape() as tape:
+            out_circuit = c(x[0], *layer.trainable_variables)
+
+        g_circuit = tape.gradient(out_circuit, layer.trainable_variables)
+
+        for i in range(len(out_layer)):
+            assert np.allclose(g_layer[i], g_circuit[i])
+
+
+@pytest.mark.usefixtures("get_circuit", "model")
+class TestKerasLayerIntegration:
+    """Integration tests for the pennylane.qnn.keras.KerasLayer class."""
+
+    @pytest.mark.parametrize("interface", qml.qnodes.decorator.ALLOWED_INTERFACES)
+    @pytest.mark.parametrize("n_qubits, output_dim", indicies_up_to(2))
+    @pytest.mark.parametrize("batch_size", [5, 10])
+    def test_train_model(self, model, batch_size, n_qubits, output_dim):
+        """Test if a model can train using the KerasLayer. The model is composed of a single
+        KerasLayer sandwiched between two Dense layers, and the dataset is simply input and output
+        vectors of zeros."""
+
+        x = np.zeros((5, n_qubits))
+        y = np.zeros((5, output_dim))
+
+        model.compile(optimizer="sgd", loss="mse")
+
+        model.fit(x, y, batch_size=batch_size, verbose=0)
+
+    @pytest.mark.parametrize("interface", qml.qnodes.decorator.ALLOWED_INTERFACES)
+    @pytest.mark.parametrize("n_qubits, output_dim", indicies_up_to(2))
+    def test_model_gradients(self, model, output_dim, n_qubits):
+        """Test if a gradient can be calculated with respect to all of the trainable variables in
+        the model"""
+        x = tf.zeros((5, n_qubits))
+        y = tf.zeros((5, output_dim))
+
+        with tf.GradientTape() as tape:
+            out = model(x)
+            loss = tf.keras.losses.mean_squared_error(out, y)
+
+        gradients = tape.gradient(loss, model.trainable_variables)
+        assert all([g.dtype == tf.keras.backend.floatx() for g in gradients])
+
+    @pytest.mark.parametrize("interface", ["tf"])
+    @pytest.mark.parametrize("n_qubits, output_dim", indicies_up_to(2))
+    def test_model_save_weights(self, model, n_qubits):
+        """Test if the model can be successfully saved and reloaded using the get_weights()
+        method"""
+        _, filename = tempfile.mkstemp()
+        prediction = model.predict(np.ones(n_qubits))
+        weights = model.get_weights()
+        model.save_weights(filename)
+        model.load_weights(filename)
+        prediction_loaded = model.predict(np.ones(n_qubits))
+        weights_loaded = model.get_weights()
+        os.remove(filename)
+
+        assert np.allclose(prediction, prediction_loaded)
+        for i, w in enumerate(weights):
+            assert np.allclose(w, weights_loaded[i])
```

Comment:
Does that mean that model saving now works with quantum keras layers?

Reply:
There are two ways to save a model in Keras (https://www.tensorflow.org/guide/keras/save_and_serialize):

- `model.save()`, which completely saves the model - including e.g. the structure, the choice of optimizer, the method of weight initialization. It should be completely self contained and does not require any setup to load other than `model.load()`.

- `model.save_weights()` saves just the weights constituting a model. It allows the model to be quickly re-initialized if one already has the code to set up things like structure.

Currently we can do the second option, which should be pretty good for now. For `model.save()`, I think we'd have to look into serializing the QNode.